#VRML_SIM R2023b utf8
# Ths is a proto for our robot for the final project
# template language: javascript


PROTO SeeingEyeRobot [
  field SFVec3f translation 0 0 0
  field SFRotation rotation 0 0 1 0
  field SFFloat bodyMass 1
  field SFString name "robot"
  field SFString controller "<generic>"
]
{
 Robot {
   translation IS translation
   rotation IS rotation
   name IS name
   controller IS controller
    children [
    DEF ROBOT_BASE Solid {
      translation 0 0 -0.0016935350243008074
      children [
        Solid {
          translation 0.28 0 -0.00999987
          rotation 0 1 0 -1.5707953071795862
          children [
            Lidar {
              translation 5.87712e-07 0 -0.05
              rotation 0 1 0 1.5708
              name "front_lidar"
            }
            Shape {
              appearance MattePaint {
                baseColor 1 1 0
              }
              geometry Capsule {
                height 0.05
                radius 0.025
              }
            }
          ]
          name "Front_Lidar_Solid"
        }
        Solid {
          translation 0.29 0 0.08
          rotation 0 1 0 -1.5707953071795862
          children [
            Shape {
              appearance MattePaint {
                baseColor 0 1 1
              }
              geometry Cylinder {
                height 0.05
                radius 0.02
              }
            }
            Camera {
              translation 0 0 -0.025
              rotation 0 1 0 1.5708
              name "front_camera"
            }
          ]
          name "Front_Camera_Solid"
        }
        Solid {
          translation 0 0 0.15
          children [
            Shape {
              appearance MattePaint {
                baseColor 0 0.333333 0
              }
              geometry Box {
                size 0.05 0.05 0.05
              }
            }
            GPS {
            }
          ]
          name "GPS"
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            axis 0 1 0
            anchor 0.2 -0.2 -0.15
          }
          device [
            RotationalMotor {
            }
            PositionSensor {
            }
          ]
          endPoint Solid {
            translation 0.2 -0.2 -0.15
            rotation 1 -1.0370632309154904e-11 1.0370807250451913e-11 1.5707996938995423
            children [
              DEF Wheel Shape {
                appearance MattePaint {
                  baseColor 0 0 0
                }
                geometry DEF Wheel_size Cylinder {
                  height 0.05
                  radius 0.08
                }
              }
            ]
            name "wheel_1"
            contactMaterial "InteriorWheelMat"
            boundingObject USE Wheel_size
            physics Physics {
            }
          }
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            axis 0 1 0
            anchor 0.2 0.2 -0.15
          }
          device [
            RotationalMotor {
            }
            PositionSensor {
            }
          ]
          endPoint Solid {
            translation 0.2 0.2 -0.15
            rotation 1 -1.0372312618623644e-11 1.0372170329234128e-11 1.5707996938995434
            children [
              USE Wheel
            ]
            name "wheel_2"
            contactMaterial "ExteriorWheelMat"
            boundingObject USE Wheel_size
            physics Physics {
            }
          }
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            axis 0 1 0
            anchor -0.2 -0.2 -0.15
          }
          device [
            RotationalMotor {
            }
            PositionSensor {
            }
          ]
          endPoint Solid {
            translation -0.2 -0.2 -0.15
            rotation 1 -1.0371633799625908e-11 1.0371747160129742e-11 1.5707996938995388
            children [
              USE Wheel
            ]
            name "wheel_3"
            contactMaterial "ExteriorWheelMat"
            boundingObject USE Wheel_size
            physics Physics {
            }
          }
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            axis 0 1 0
            anchor -0.2 0.2 -0.15
          }
          device [
            RotationalMotor {
            }
            PositionSensor {
            }
          ]
          endPoint Solid {
            translation -0.2 0.2 -0.15
            rotation 1 -1.0372160841997104e-11 1.037223112421577e-11 1.5707996938995399
            children [
              USE Wheel
            ]
            name "wheel_4"
            contactMaterial "InteriorWheelMat"
            boundingObject USE Wheel_size
            physics Physics {
            }
          }
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            axis 0 0 1
            anchor -0.12 0 0.2
          }
          device [
            PositionSensor {
            }
            RotationalMotor {
              name "gimbal_j1_motor"
            }
          ]
          endPoint Solid {
            translation -0.12 0 0.2
            rotation 9.024993567198802e-17 1 0 4.692706653739929e-06
            children [
              HingeJoint {
                jointParameters HingeJointParameters {
                  axis 0 1 0
                  anchor 0 0 0.08
                  minStop -0.52
                  maxStop 0.52
                }
                device [
                  RotationalMotor {
                    name "gimbal_j2_motor"
                  }
                  PositionSensor {
                  }
                ]
                endPoint Solid {
                  translation 0 0 0.08
                  rotation 1 0 0 1.5707999999999998
                  children [
                    Solid {
                      rotation 1 0 0 -1.5707953071795862
                      children [
                        Camera {
                          translation -0.036 0 0
                          rotation 0 0 1 3.141592653589793
                          name "RGB_camera"
                          recognitionColors [
                            0 0 0
                          ]
                          recognition Recognition {
                          }
                        }
                        RangeFinder {
                          translation -0.035 0 0
                          rotation 0 0 1 3.141592653589793
                          name "range_finder"
                          width 640
                          height 480
                          minRange 0.02
                          maxRange 4
                        }
                        Shape {
                          appearance MattePaint {
                            baseColor 0 0 0
                          }
                          geometry Box {
                            size 0.07 0.18 0.04
                          }
                        }
                      ]
                      name "Depth_Camera"
                    }
                    Shape {
                      appearance MattePaint {
                        baseColor 0 0.666667 1
                      }
                      geometry Cylinder {
                        height 0.2
                        radius 0.03
                      }
                    }
                  ]
                  name "Gimbal_Joint_2"
                }
              }
              Shape {
                appearance MattePaint {
                  baseColor 0.333333 1 1
                }
                geometry Cylinder {
                  height 0.1
                  radius 0.04
                }
              }
            ]
            name "Gimbal_Joint_1"
          }
        }
        Shape {
          appearance MattePaint {
            baseColor 0.333333 0 1
          }
          geometry DEF BASE Box {
            size 0.6 0.3 0.3
          }
        }
      ]
      name "Gimbal"
      boundingObject USE BASE
      physics Physics {
      density -1
      mass IS bodyMass
      }
    }
  ]
  controller IS controller
}
}



